ui/src/features/layer/new/controller/layer.controller.ts

```diff
@@ -0,0 +1,276 @@
+
+  abstract get type(): LayerType;
+
+  /**
+   * Updates the controller's current layer data,
+   * adjusting the Cesium viewer where needed.
+   *
+   * @param layer The updated layer data.
+   */
+  update(layer: T): void {
+    // Update the local layer data.
+    // Note that we do this *before* checking for changes, as change detection is handled via `watchedValues`.
+    this._layer = layer;
+
+    // Check if there are any changes between the current and the new dataset.
+    this.reactToChanges();
+
+    // If the changes require the current Cesium layer to be removed and a new one to be added,
+    // we have to re-run `addToViewer`.
+    if (this.hasRequestedReinitialization) {
+      this.addToViewer();
+    }
+    this.hasRequestedReinitialization = false;
+
+    // Apply any adjustments to the current layer.
+    // Note that we do this even if we have reinitialized the layer,
+    // as there may be adjustments that are not handled by `addToViewer` itself.
+    for (const change of this.requestedChanges) {
+      change();
+    }
+
+    // Reset the temporary watch fields.
+    this.currentWatchIndex = 0;
+    this.requestedChanges = [];
+  }
+
+  /**
+   * Remove the layer from the viewer and destroy the controller's owned data.
+   */
+  remove(): void {
+    this.removeFromViewer();
+  }
+
+  /**
+   * Zooms the viewer to the layer's location.
+   */
+  abstract zoomIntoView(): void;
+
+  /**
+   * Moves the layer to the top of the map.
+   */
+  abstract moveToTop(): void;
+
+  /**
+   * Checks if the Cesium layer needs to be adjusted based on the current layer data.
+   *
+   * Note that this method should mostly consist of calls to {@link watch},
+   * with no conditional or other side effects.
+   *
+   * @protected
+   */
+  protected abstract reactToChanges(): void;
+
+  /**
+   * Adds the layer to the viewer.
+   * If the layer has already been added, this instead replaces the current Cesium layer with a new one.
+   *
+   * @protected
+   */
+  protected abstract addToViewer(): void;
+
+  /**
+   * Removes the Cesium layer from the viewer, fully destroying it in the process.
+   *
+   * @protected
+   */
+  protected abstract removeFromViewer(): void;
+
+  /**
+   * Compares a value to its previous version.
+   *
+   * If the value has changed, `action` will be called.
+   * If `action` doesn't exist, a reinitializaton is requested instead.
+   *
+   * Note that when a value is checked for the first time (i.e. there is no preceding version),
+   * this is essentially a no-op, called only to register the value for later comparison.
+   *
+   * @param value The update value.
+   * @param action An action to execute when the value has changed. This is assumed to adjust the Cesium layer.
+   * @protected
+   */
+  protected watch<T>(value: T, action?: (value: T) => void): void {
+    if (!this.isInitialized) {
+      this.watchedValues.push(value);
+    }
+    const lastValue = this.watchedValues[this.currentWatchIndex];
```

Comment:
Shouldn't this be something like this:

```suggestion
    const lastValue = this.watchedValues[this.currentWatchIndex - 1];
```

Because you are updating the currentWatchINDex two lines later and then there will never be a value at that index. Or am i missing something?